File: src/converter/plugins/convertVariableStatement.ts
import ts from "typescript";
import {createSimplePlugin} from "../plugin.js";
import {CheckCoverageService, checkCoverageServiceKey} from "./CheckCoveragePlugin.js";

export const convertVariableStatement = createSimplePlugin((node, context, render) => {
    if (!ts.isVariableStatement(node)) return null

    const checkCoverageService = context.lookupService<CheckCoverageService>(checkCoverageServiceKey)
    checkCoverageService?.cover(node)
    checkCoverageService?.cover(node.declarationList)

    const exportModifier = node.modifiers?.find(it => it.kind === ts.SyntaxKind.ExportKeyword)
    exportModifier && checkCoverageService?.cover(exportModifier)

    return node.declarationList.declarations
        .map(declaration => render(declaration))
        .join("\n")
})

File: gradle-plugin/karakum/src/jsMain/kotlin/io/github/sgrishchenko/karakum/extension/plugins/convertVariableStatement.kt
package io.github.sgrishchenko.karakum.extension.plugins

import io.github.sgrishchenko.karakum.extension.createSimplePlugin
import typescript.Node
import typescript.SyntaxKind
import typescript.asArray
import typescript.isVariableStatement

val convertVariableStatement = createSimplePlugin plugin@{ node: Node, context, render ->
    if (!isVariableStatement(node)) return@plugin null

    val checkCoverageService = context.lookupService<CheckCoverageService>(checkCoverageServiceKey)
    checkCoverageService?.cover(node)
    checkCoverageService?.cover(node.declarationList)

    val exportModifier = node.modifiers?.asArray()?.find { it.kind == SyntaxKind.ExportKeyword }
    exportModifier?.let { checkCoverageService?.cover(exportModifier) }

    node.declarationList.declarations.asArray().joinToString(separator = "\n") { render(it) }
}
